grails-app/services/org/pih/warehouse/inventory/CycleCountService.groovy

```diff
@@ -390,33 +374,31 @@ class CycleCountService {
         return cycleCounts.collect { CycleCountDto.toDto(it) }
     }
 
+    /**
+     * Submits the (re)count as it is in its current state.
+     */
     CycleCountDto submitCount(CycleCountSubmitCountCommand command) {
-        command.cycleCount.cycleCountItems.each { CycleCountItem cycleCountItem ->
-            updateCycleCountItemForSubmit(cycleCountItem, command.refreshQuantityOnHand, command.failOnOutdatedQuantity)
+        CycleCount cycleCount = command.cycleCount
+
+        if (command.refreshQuantityOnHand) {
+            boolean hasQuantityChanged = cycleCountProductAvailabilityService.refreshProductAvailability(cycleCount)
```

Comment:
we now do the QoH refresh outside of the cycleCountItems loop below. We can't just loop the items of the count because there might be new items created that the count is not aware of.

Reply:
suggestion: It would be cool to separate the "detect changes" from the "apply changes" in case we want show the user what's about to happen. Detect changes could return a list of changes that could be displayed to the user.

Reply:
yeah I was actually considering doing something like this originally but it got kinda complex (I needed three lists, delete, create, update). I'll give it another think through again though